### ISSUE TITLE
Add a way to disable jscs tests when running localhost:4200/tests

### ISSUE THREAD

**Issue by toddjordan** [dockyard/ember-suave#62]
My team is a user of ember-suave (thanks for this, its great), and we'd like the ability to disable jscs tests in the testem UI.   Our editors are good at enforcing jscs and while we like the CI capability, we'd like to have the option to only run our unit tests through testem in our dev workflow.  Currently there is an option to "Disable JSHint" out of the box with ember cli.  We'd additionally like an option when you install suave to get a "Disable JSCS" checkbox.

**Comment by toddjordan** [comment]
I'd be happy to create a PR for this.  I'm currently looking through code to see what other addons are doing to add these options.  Another alternative might be to hook it to the "Hide JSHint" option, but I'm unsure if there's a use case for wanting one and not that other.

**Comment by toddjordan** [comment]
This seems to be only an issue when I run ```ember s``` and use localhost:4200/tests.  Running ember test --server is skipping jscs.  In looking at code and what ember-suave and broccoli-jscs does, I'm just going to look at getting /tests to skip jscs test generation.

**Comment by toddjordan** [comment]
Alright... After some more investigation yesterday I figured out the following:  

Its only an issue when I run ```ember s``` because I don't think lintTree gets run on ```ember test```, thus no suave.  

Secondly, I backed off of adding qunit customization to the added at least for now, since it doesn't currently have that dependency, and I'd be fine with just always disabling the test generation, which you can theoretically do by setting "disableTestGeneration" on my jscs options.  

Unfortunately, setting "disableTestGeneration" is mucking with babel, destroying test execution.  I found an issue against broccoli-jscs that describes this, and points to it being an ember-cli issue: https://github.com/kellyselden/broccoli-jscs/issues/31, and the ember-cli issue: https://github.com/ember-cli/ember-cli/issues/4340,  which I think fixes it when the next ember-cli comes out (fixed 5 days ago).

Anyways, I can close this issue or leave it open for a potential future enhancement to augment qunit with a new checkbox, which would prevent others from going down this road, or at least document passing disableTestGenerator in the readme.  For me, an optimal setup would be to not generate the tests when running the testem UI, but to do so during CI, which would suggest implementing the checkbox.

**Comment by toddjordan** [comment]
ok, back on the case (unless someone else has already picked up).  We are now needing the tests when we do a push (via pre-push hook), but we don't want to see 1000 jscs tests in our testem UI, so there is motivation for the UI update and not a flat out test generation disable... Plus I heard the issue mentioned on Ember.Land ™ so I feel like someone else cares ;-)  I plan to go about the fix by including qunit as a dependency to the addon, and then passing the config to show the extra option from there.  I'm new to the addon ecosystem so if anyone knows a better way lmk.

**Comment by brzpegasus** [comment]
Sorry! Been really busy, and this issue completely slipped my mind. I'll make some time to respond properly soon, hopefully tonight.

**Comment by toddjordan** [comment]
No problem!  I definitely relate to busyness :-) Like I said I'm happy to contribute back, but some validation on the approach would be great.  I see qunit-notifications as an example of adding a new checkbox from an addon, but I noticed that actually is directly manipulating the testem dom to add it.  
https://github.com/dockyard/qunit-notifications/blob/master/index.js

I see in other tools, such as blanket where they add config to QUnit like:
https://github.com/alex-seville/blanket/blob/master/src/qunit/qunit.js

```javascript
QUnit.config.urlConfig.push({
            id: "coverage",
            label: "Enable coverage",
            tooltip: "Enable code coverage."
        });
```

My thought is we can do something similar to blanket, and that's the road I'm starting down...

**Comment by brzpegasus** [comment]
Thanks for pushing this. I definitely think there is value in having a checkbox for disabling JSCS tests, but it may not be very straightforward.

First, let me clear up a few things:

* Testem is only involved when you run `ember test` (runs testem in CI mode), or `ember test --server` (runs testem in dev mode).
* When you run `ember serve` and navigate to `http://localhost:4200/tests`, it's really Ember CLI's default Express [server](https://github.com/ember-cli/ember-cli/blob/v1.13.8/lib/tasks/server/express-server.js) that is serving the test page via some special [test middleware](https://github.com/ember-cli/ember-cli/blob/v1.13.8/lib/tasks/server/middleware/tests-server/index.js#L18-L36).

Now, to the actual issue. In order to prevent JSCS tests from loading, you actually have to do two things:

* Push a "Disable JSCS" option to [`QUnit.config.urlConfig`](https://api.qunitjs.com/QUnit.config/). QUnit will add a checkbox for each item specified in that array.
* Update Ember CLI's [`test-loader`](https://github.com/ember-cli/ember-cli-qunit/blob/v1.0.1/vendor/ember-cli-qunit/test-loader.js#L5-L7) to skip modules ending in `.jscs` if that checkbox is selected.

That second part is tricky, because there are no extension points, from what I see, for you to customize the loader, short of overwriting `TestLoader.prototype.shouldLoadModule` after `test-loader.js` executes, but before [`TestLoader.load()`](https://github.com/ember-cli/ember-cli-qunit/blob/v1.0.1/vendor/ember-cli-qunit/test-loader.js#L20) is called. That's obviously really hacky.

I feel it might be best to either make the test loader more extensible so we can do these sorts of things, or update the TestLoader itself to expose the option (but JSCS isn't included by default, unlike JSHint, so I'm not sure that would pass). Either way, some changes in [`ember-cli-qunit`](https://github.com/ember-cli/ember-cli-qunit) might be needed to support this.

**Comment by toddjordan** [comment]
Thanks for the pointers into the test loaders.  I'll take a look at the source there and try to see what we can do.  Another thing I'm trying to figure out where the right place is to get a reference to QUnit and push the config from within the ember addon.

**Comment by brzpegasus** [comment]
It's all part of ember-cli-qunit. As you can see [here](https://github.com/ember-cli/ember-cli-qunit/blob/c775d09ee4c5122e6b9c9e7fb22a734ac3628e67/index.js#L46-L65), `qunit.js`, `qunit-configuration.js`, and `test-loader.js` get concatenated (in that order) into the single, final `test-support.js` file that you can find in your `dist/assets` folder when you build.

You have access to `QUnit` globally once that file loads, but even if you were to inject additional code that modifies the config afterwards, that still won't help you override [this](https://github.com/ember-cli/ember-cli-qunit/blob/v1.0.1/vendor/ember-cli-qunit/test-loader.js#L5-L7).

/cc @rwjblue - Do you think ember-cli-qunit should be JSCS-aware?

**Comment by rwjblue** [comment]
No, but I do think it should be easy for us to add additional modules to the 'should load module' callback. 

I also hate that we have to override the prototype method in ember-cli-quint. 

I'll try to comment more fully shortly with a path forward...

**Comment by rwjblue** [comment]
https://github.com/ember-cli/ember-cli-test-loader/pull/14
https://github.com/ember-cli/ember-cli-qunit/pull/77

**Comment by brzpegasus** [comment]
@rwjblue Thanks!! That was a really fast turnaround.

**Comment by toddjordan** [comment]
ok, @brzpegasus so what I'll look at doing is adding a vendor/ember-suave/test-loader.js to add the new callbacks to test-loader similar to @rwjblue did in the ember-cli-qunit PR, then I'll update the index.js include hook to concat that to test-support.js.  In the same way I'll also add a script to push the config to QUnit that will add the checkbox.  Does this sound like a reasonable approach?

**Comment by toddjordan** [comment]
Got it working, created a pull request...
https://github.com/dockyard/ember-suave/pull/63/files

**Comment by brzpegasus** [comment]
@toddjordan Awesome, thanks!

**Comment by toddjordan** [comment]
Tried it out with the latest versions of ember-cli-test-loader and ember-cli-qunit and works!  Will putting those versions in the addon package.js/bower.json force apps to install them?

**Comment by brzpegasus** [comment]
No, the consuming app would have to update their `ember-cli-qunit` and `ember-cli-test-loader` versions. I can make a note of that when releasing though.

**Comment by brzpegasus** [comment]
Closed by #63.
